[ADAM-808] build an assembly cli jar with maven shade plugin

heuermh opened the pull request:
Another approach to #808.

Still in progress (e.g. I'm not sure how ```adam-pyspark``` works), thought I'd should push it early for review/discussion.

Will depend on a fix for #812.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/889/
Test PASSed.

massie commented:
Thanks for this PR, Michael. 

We originally created an uber jar for distribution but ran into file limitations with zip/jar files (see http://stackoverflow.com/questions/9616250/what-is-the-maximum-number-of-files-per-jar). Now that we've have Spark and Hadoop as `provided` dependencies, the number of files in our jar file should have dropped significantly.

How much head room do we have? How many files are in this uber jar, e.g. what does `jar -tvf adam-uber.jar | wc -l` output?

fnothaft commented:
I think it'd make sense to move back to the uberjar if we've got enough head room.

fnothaft commented on adam-cli/pom.xml:
```diff
@@ -14,22 +14,35 @@
+                <exclude>META-INF/*.RSA</exclude>
+              </excludes>
+            </filter>
+          </filters>
```
OOC, I went back to the ol' days before we moved to appassembler and saw a few more exclusions (see at https://github.com/bigdatagenomics/adam/blob/adam-parent-0.6.0/adam-cli/pom.xml#L43). Might be worth doing a diff of the old `maven-shade-plugin` section versus the new one.

heuermh replied:
Yep, and there are quite a few other transformers that e.g. the Spark project uses, will have to look further

massie replied:
@heuermh Did you see any changes that needed to by made to the PR when you looked at the old `maven-shade-plugin` xml?

This RP looks good to me but, since it's a big change (and I've hit the merge button too quickly a few times lately), I wanted to check.

fnothaft replied:
I'll build this and run on the cluster and see how it goes.

massie replied:
Thanks, Frank.

heuermh replied:
This is definitely not ready to go yet.

It still needs
 - removal of ```compute-adam-classpath.sh``` and ```compute-adam-jars.sh```; I haven't done this yet because I'm not sure how ```adam-pyspark``` is supposed to work
 - a careful look at maven-shade-plugin excludes/transformer configuration as mentioned above
 - decision on whether to continue using cli module as the assembly artifact or create a separate assembly module as in the Spark build
 - cleanup of changes to ```adam-shell.sh``` and ```adam-submit.sh``` per review notes
 - test assembly artifact on YARN cluster, to confirm issues such as #663 don't resurface
 - test assembly artifact in build for downstream applications

fnothaft replied:
> removal of compute-adam-classpath.sh and compute-adam-jars.sh; I haven't done this yet because I'm not sure how adam-pyspark is supposed to work

I think it is OK to remove both of those, and possibly `adam-pyspark`? CC @laserson 

> decision on whether to continue using cli module as the assembly artifact or create a separate assembly module as in the Spark build

What exactly would the difference be here?

> test assembly artifact on YARN cluster, to confirm issues such as #663 don't resurface

This seems to be OK.

> test assembly artifact in build for downstream applications

I don't think this change should impact downstream applications, except possibly plugins. Are plugins your concern?

fnothaft replied:
`adam-submit` and `adam-shell` are both OK on YARN (the issue seen in #663 does _not_ reproduce).

heuermh replied:
>> test assembly artifact in build for downstream applications

> I don't think this change should impact downstream applications, except possibly plugins. Are plugins your concern?

avocado, rice, qc-metrics, etc. all use classpath-determining scripts similar to those being removed here in their ```bin``` directories.

fnothaft replied:
> avocado, rice, qc-metrics, etc. all use classpath-determining scripts similar to those being removed here in their bin directories.

Oh, sure, but this change doesn't change the adam-{core,apis,cli} artifact that is published to Maven Central.
We can change the packaging in the downstream projects at a later date.

heuermh replied:
sorry, I've lost the threading for this comment.

I went back to the old shade plugin configuration and don't see those exclusions as being necessary any more.  I don't see ```reference.conf``` either, so I'm not sure what that transformer line was for.

fnothaft commented on bin/adam-shell:
```diff
@@ -28,8 +26,34 @@ if [[ -z $@ && -n "$ADAM_OPTS" ]]; then
 
+# Find ADAM cli assembly jar
+ADAM_CLI_JAR=
+if [ -f "$SCRIPT_DIR/repo" ]; then
```
Why are we checking for a `/repo` directory here? Is this for backwards compatibility with appassembler?

heuermh replied:
I updated the distribution assembly.xml to copy the adam-cli jar to repo, so this check is to see if the user is trying to run from the dist.  The Spark build uses a separate assembly module and has a more involved find (in the ```spark-class``` script), we may wish to crib from there.

fnothaft replied:
Ah, OK! Got it.

fnothaft commented on bin/adam-submit:
```diff
@@ -52,15 +52,29 @@ fi
+# Find ADAM cli assembly jar
+ADAM_CLI_JAR=
+if [ -f "$SCRIPT_DIR/repo" ]; then
+  ASSEMBLY_DIR="$SCRIPT_DIR/repo"
```
Same question as above here.

fnothaft commented on bin/adam-submit:
```diff
@@ -52,15 +52,29 @@ fi
-# Split out the CLI jar, since it will be passed to Spark as the "primary resource".
-ADAM_CLI_JAR=${ADAM_JARS##*,}
-ADAM_JARS=$(echo "$ADAM_JARS" | rev | cut -d',' -f2- | rev)
+num_jars="$(ls -1 "$ASSEMBLY_DIR" | grep "^adam-cli.*\.jar$" | wc -l)"
```
I'd find it preferable to do:

```
jars=({ASSEMBLY_DIR}/adam-cli*.jar)
num_jars=${#jars[*]}
```

Or something of the sort.

heuermh replied:
This bit also needs to be more clever, avoiding other artifacts such as adam-cli_...-javadoc.jar
```bash
$ ./bin/adam-shell -h
Found multiple ADAM cli assembly jars in /Users/heuermh/working/adam/adam-cli/target:
adam-cli_2.10-0.17.2-SNAPSHOT-javadoc.jar
adam-cli_2.10-0.17.2-SNAPSHOT.jar
Please remove all but one jar.
```

heuermh commented:
```bash
$ jar -tvf adam-cli/target/adam-cli_2.10-0.17.2-SNAPSHOT.jar | wc -l
   37210
```
and the limitation was only with JDK6 and earlier, it shouldn't be a problem with JDK7 and later.

fnothaft commented:
If we are at half of the limit, then I would go with this.

> and the limitation was only with JDK6 and earlier, it shouldn't be a problem with JDK7 and later.

Is ZIP64 the default in JDK7 and onwards?

massie commented:
Agree. We have plenty of headroom. Zip64 is supported from JDK7 onward. It appears reading online that `zip` is default and, if there are >65535 files, that `zip64` is used. I guess they did that for backward compatibility (since pre-JDK7 can't read zip64).

massie commented:
@heuermh You mentioned that the following needs to be done. How's it going?

- [ ] removal of compute-adam-classpath.sh and compute-adam-jars.sh; I haven't done this yet because I'm not sure how adam-pyspark is supposed to work
- [ ] a careful look at maven-shade-plugin excludes/transformer configuration as mentioned above decision on whether to continue using cli module as the assembly artifact or create a separate assembly module as in the Spark build
- [ ] cleanup of changes to adam-shell.sh and adam-submit.sh per review notes
- [ ] test assembly artifact on YARN cluster, to confirm issues such as #663 don't resurface
- [ ] test assembly artifact in build for downstream applications

@fnothaft Are you using this on the cluster with success?

heuermh commented:
Pushing after rebase with additional changes:  removed ```adam-pyspark``` and ```compute-*``` scripts; removed ```NOTICE.txt``` as cofoja is no longer a dependency; and cleaned up per review above.

I couldn't get this to do the right thing, even with some twiddling
```bash
jars=({ASSEMBLY_DIR}/adam-cli*.jar)
num_jars=${#jars[*]}
```

fnothaft commented:
> I couldn't get this to do the right thing, even with some twiddling

What (wrong thing) does it do instead?

heuermh commented:
> What (wrong thing) does it do instead?

If I change it to
```bash
jars=({$ASSEMBLY_DIR}/adam-cli*.jar)
```
then bash doesn't complain but
```bash
num_jars=0
```
when it shouldn't be.  Maybe an OSX thing?

fnothaft commented:
I will check this out on my end.

fnothaft commented:
Also, the above RE: `num_jars` looks fine by me as it is.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/947/
Test PASSed.

fnothaft commented:
@heuermh since this is a big build change, can you rebase this on ToT? Once it passes @AmplabJenkins, I will merge.

heuermh commented:
I might want to remove ```<mainClass>org.bdgenomics.adam.cli.ADAMMain</mainClass>``` as well.  Per issue #808 running the adam cli jar will fail.  WDYT?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/955/
Test PASSed.

fnothaft commented:
@heuermh that sounds like a good idea to me. Can you make that change and push it?

heuermh commented:
ok

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/956/
Test PASSed.

fnothaft commented:
Thanks @heuermh! I've verified this on the cluster, and have merged it as https://github.com/bigdatagenomics/adam/commit/f39de5743e2d995db0eeb22cd324e9dd87596817.

heuermh commented:
Thanks!
